From 8e34da261cbb04e91088ea94b8742b4d38b34e24 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Thu, 28 Jun 2018 19:53:27 +0200
Subject: [PATCH 1/1] lib: div64: fix typeo in include/div64.h

%s/reminder/remainder/

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 include/div64.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/div64.h b/include/div64.h
index cecb2322bc..76563ef978 100644
--- a/include/div64.h
+++ b/include/div64.h
@@ -232,7 +232,7 @@ extern uint32_t __div64_32(uint64_t *dividend, uint32_t divisor);
 #endif /* BITS_PER_LONG */
 
 /* Wrapper for do_div(). Doesn't modify dividend and returns
- * the result, not reminder.
+ * the result, not remainder.
  */
 static inline uint64_t lldiv(uint64_t dividend, uint32_t divisor)
 {
-- 
2.18.0

