From d4b5c1563c09e48d29afeb246a8794768596e06b Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 30 Dec 2018 12:58:33 +0100
Subject: [PATCH 1/1] doc: README.commands: fix typo

%s/commmand/command/

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 doc/README.commands | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/doc/README.commands b/doc/README.commands
index 0ccadae0b7..e03eb44187 100644
--- a/doc/README.commands
+++ b/doc/README.commands
@@ -67,7 +67,7 @@ This table has to be evaluated in the command function of the main command, e.g.
 Command function
 ----------------
 
-The commmand function pointer has to be of type
+The command function pointer has to be of type
 int (*cmd)(struct cmd_tbl_s *cmdtp, int flag, int argc, const char *argv[]);
 
 cmdtp:		Table entry describing the command (see above).
-- 
2.19.2

