From 99922654957b69e15e93c88b906980f8e2c4d059 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Fri, 30 Oct 2020 06:10:02 +0100
Subject: [PATCH 1/1] env: typo enougth

%s/enougth/enough/

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 env/nowhere.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/env/nowhere.c b/env/nowhere.c
index d33fdf27d0..d43a2d6b51 100644
--- a/env/nowhere.c
+++ b/env/nowhere.c
@@ -30,7 +30,7 @@ static int env_nowhere_init(void)
 static int env_nowhere_load(void)
 {
 	/*
-	 * for SPL, set env_valid = ENV_INVALID is enougth as env_get_char()
+	 * for SPL, set env_valid = ENV_INVALID is enough as env_get_char()
 	 * return the default env if env_get is used
 	 * and SPL don't used env_import to reduce its size
 	 * For U-Boot proper, import the default environment to allow reload.
-- 
2.28.0

