From 4ba511581ad7e293d577c9ee1c9accb976dc9cf2 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 22 Mar 2020 19:12:02 +0100
Subject: [PATCH 1/1] test: typo decompression

%s/decopmression/decompression/

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 test/compression.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/compression.c b/test/compression.c
index cf040d7c86..63f929fa4b 100644
--- a/test/compression.c
+++ b/test/compression.c
@@ -451,7 +451,7 @@ static int compress_using_none(struct unit_test_state *uts,
 }
 
 /**
- * run_bootm_test() - Run tests on the bootm decopmression function
+ * run_bootm_test() - Run tests on the bootm decompression function
  *
  * @comp_type:	Compression type to test
  * @compress:	Our function to compress data
-- 
2.25.1

